![alt text](https://github.com/MattMason95/graphGSEA/blob/94324aac74576dab7605851f5da81bfb620f3af7/graphGSEA/graphGSEA.png)

## Table of contents
- [About](#about)
- [Tutorial](#tutorial)

## About <a name = "about"></a>
Geneset enrichment analysis (GSEA) is a mainstay of RNA-sequencing analysis workflows. 
Functional enrichment analyses afford exceptional utility for both unbiased data exploration, and hypothesis driven approaches. However, due to the redundancy of query genesets utilised from the Gene Ontology database, GSEA can output many enriched genesets that overlap significantly in both semantic annotation and gene composition. 

graphGSEA seeks to overcome this redundancy by building relational networks of genesets and analysing subgraphs of related genesets. In this context, genesets are represented as nodes. Nodes are evaluated in a pair-wise manner to compute Jaccard indices of their overlapping genes. Edges are drawn between nodes that have a Jaccard index of > 0.5 (i.e. more than 50% of their genes are shared), under the assumption that genesets sharing the majority of their genes will be involved in similar functions. The stringency/leniency of this threshold can be selected at will.

This package contains several utilities for the parsing of data from the Windows installation of GSEA, the generation of relational network graphs from significantly enriched genesets, the identification of interconnected subgraphs, and a tokenisation of the geneset annotations that populate subgraphs of interest. 
By evaluating subgraphs of interrelated nodes, instead of individual genesets, the redundancy of the analysis is reduced. 

## Tutorial <a name = "tutorial"></a>
Import the parseGSEA function from the parsing utility folder. Use the fileAcccessor() function to extract data from the designated GSEA experiment directory containing the EDB folder.
This will return a networkResults object, which stores the network label, the network dataframe, network metadata, network statistics, and a gene lookup table.

```python

from parsing import parseGSEA

outputs = parseGSEA(filepath='./Example_data').fileAccessor(jaccardFilter=0.5,statThreshold=0.1)


# ~~~~~~ parseGSEA ~~~~~~

# Retrieving data from Example_data.
# EDB folder found at:Example_data\GSEA_2m/edb.
# Parsing data at destinations.
# Parsing: Example_data\GSEA_2m/edb
# PARSED - .gmt file.
# PARSED - .edb file.
# PARSED - Significant genesets.
# Computing Jaccard indices.
# 142311it [00:01, 89845.45it/s]

# ~~~~~~ </parseGSEA> ~~~~~~
```
To analyse the composition of the relational network graph, import the analyseGSEA from the analysis utility folder.
The networkResults object generated by parseGSEA can be passed to the clusterInformation() function, which will return an analysisResulst object, comprising cluster-wise semantic enrichments, geneset compositions, and mean normalised enrichment scores.

```python
from parsing import parseGSEA
from analysis import analyseGSEA

outputs = parseGSEA(filepath='./Example_data').fileAccessor(jaccardFilter=0.5,statThreshold=0.1)
results = analyseGSEA().clusterInformation(outputs)

# ~~~~~~ parseGSEA ~~~~~~

# Retrieving data from Example_data.
# EDB folder found at: Example_data\GSEA_2m/edb.
# Parsing data at destinations.
# Parsing: Example_data\GSEA_2m/edb
# PARSED - .gmt file.
# PARSED - .edb file.
# PARSED - Significant genesets.
# Computing Jaccard indices.
# 142311it [00:01, 89845.45it/s]
# Parsing Complete.

# ~~~~~~ </parseGSEA> ~~~~~~
# ~~~~~~ analyseGSEA ~~~~~~
 
# Creating instance of graph network.
# Extracting subgraphs.
# Calculating relative enrichment of geneset annotations.
# Calculating mean normalised enrichment scores.
# Analysis complete.

# ~~~~~~ </analyseGSEA> ~~~~~~
```
Elements of the analysisResults object can be accessed accordingly. 

```python
## The contents of the analysisResults object are as follows:
import pandas as pd 
print('Semantic enrichment:\n',pd.Series(results.enrichment).head(5),'\n')
print('Cluster geneset composition:\n',pd.Series(results.clusters).head(5),'\n')
print('Cluster average normalised enrichment score:\n',pd.Series(results.clusterNES).head(5),'\n')

# Semantic enrichment:
#  2     {'PROCESSING': 44.5, 'OF': 0.45408163265306123...
# 44    {'NADH': 89.0, 'DEHYDROGENASE': 29.66666666666...
# 68    {'VIRAL': 26.7, 'PROCESS': 4.271999999999999, ...
# 80    {'MITOTIC': 14.239999999999998, 'CELL': 12.714...
# 82    {'AMPLIFICATION': 41.07692307692307, 'OF': 1.3...
# dtype: object 

# Cluster geneset composition:
#  2     {REACTOME_PROCESSING_OF_CAPPED_INTRON_CONTAINI...
# 44    {GOCC_NADH_DEHYDROGENASE_COMPLEX, GOBP_PROTON_...
# 68    {GOBP_VIRAL_PROCESS, GOBP_NEGATIVE_REGULATION_...
# 80    {GOBP_MITOTIC_CELL_CYCLE_PHASE_TRANSITION, GOB...
# 82    {REACTOME_AMPLIFICATION_OF_SIGNAL_FROM_THE_KIN...
# dtype: object 

# Cluster average normalised enrichment score:
#  2    -2.037037
# 44    1.748983
# 68    2.102373
# 80    1.721867
# 82    1.884723
# dtype: float64 
```

To plot the relational network graphs, graphGSEA makes use of the NetworkX package.
Import the plotGSEA function from the plotting utility folder. Pass the outputs from parseGSEA to the iterativePlotting() function to plot network graphs from the parsed GSEA data. 

```python
from parsing import parseGSEA
from plotting import plotGSEA

outputs = parseGSEA(filepath='./Example_data').fileAccessor(jaccardFilter=0.5,statThreshold=0.1)
plotGSEA.iterativePlotting(outputs,save=False)
```

![alt text](https://github.com/MattMason95/graphGSEA/blob/0f615d395f0c10978e4715f787159687d8333d9b/graphGSEA/GSEA_7m.png)
